### 공기의 질을 측정한 데이터
1. airquanlity 데이터에 대해서 결측치가 가장 많은 변수를 찾아서 해당 결측치를 0으로 대체하고 결측치를 제외한 평균과 0으로 대치한 후 평균의 차이를 구하시오
2. wind 변수에 대해서 Min-Max 정규화를 수행 한 후 평균 값과 Z 정규화를 수행 한 후 평균 값의 차이를 구하시오
3. 월별 평균 기온을 구하시오

In [7]:
import pandas as pd
import numpy as np

# csv 파일이 위치한 디렉토리 입력
df = pd.read_csv('data/airquality.csv')
df.head()

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

In [8]:
## info() 함수를 이용하여 결측치의 갯수를 파악
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Ozone    116 non-null    float64
 1   Solar.R  146 non-null    float64
 2   Wind     153 non-null    float64
 3   Temp     153 non-null    int64  
 4   Month    153 non-null    int64  
 5   Day      153 non-null    int64  
dtypes: float64(3), int64(3)
memory usage: 7.3 KB


In [9]:
# Ozone 컬럼의 결측치를 0으로 대체
# 대체 이전과 이후의 평균을 출력
print(df["Ozone"].isnull().sum())         # Ozone 컬럼의 결측치 개수 출력
Ozone_mean = df["Ozone"].mean()           # Ozone 컬럼의 평균

df["Ozone"].fillna(0, inplace=True)       # Ozone 컬럼의 결측치를 평균으로 대체

print(df["Ozone"].isnull().sum())         # 대체 후 Ozone 컬럼의 결측치 개수 출력
Ozone_mean2 = df["Ozone"].mean()          # 체치 후 Ozone 컬럼의 평균

print(Ozone_mean)
print(Ozone_mean2)

37
0
42.12931034482759
31.941176470588236


In [10]:
## 두 평균 값의 차이 출력
print(Ozone_mean-Ozone_mean2)

10.188133874239352


In [12]:
# Min-Max 표준화, Z 표준화 변환뒤 컬럼 추가
# 소수점 2자리까지 표시
Min = np.min(df["Wind"])
Max = np.max(df["Wind"])
df["min_max"] = round((df["Wind"] - Min)/(Max - Min), 2)

Mean = np.mean(df["Wind"])
Std = np.std(df["Wind"])
df["z"] = round((df["Wind"] - Mean)/Std, 2)

## Min-Max 표준화는 연속형 변수의 값을 0과 1사이의 값으로 변환. 스케일이 다른 두 변수를 Min-Max 변환하면 상호간에 비교가 가능
## z 표준화는 변수값에서 평균을 뺀 후 표준편차로 나누는 것으로, 변환 후 데이터의 평균은 0, 표준편차는 1

In [13]:
df.describe()

Ozone     Solar.R        Wind        Temp       Month         Day  \
count  153.000000  146.000000  153.000000  153.000000  153.000000  153.000000   
mean    31.941176  185.931507    9.957516   77.882353    6.993464   15.803922   
std     33.924497   90.058422    3.523001    9.465270    1.416522    8.864520   
min      0.000000    7.000000    1.700000   56.000000    5.000000    1.000000   
25%      4.000000  115.750000    7.400000   72.000000    6.000000    8.000000   
50%     21.000000  205.000000    9.700000   79.000000    7.000000   16.000000   
75%     46.000000  258.750000   11.500000   85.000000    8.000000   23.000000   
max    168.000000  334.000000   20.700000   97.000000    9.000000   31.000000   

          min_max           z  
count  153.000000  153.000000  
mean     0.433333   -0.000065  
std      0.185696    1.003679  
min      0.000000   -2.350000  
25%      0.300000   -0.730000  
50%      0.420000   -0.070000  
75%      0.520000    0.440000  
max      1.000000    3.060000

- Min-Max 표준화된 Wind 컬럼의 평균은 0.433, Z 표준화 된 컬럼의 평균은 0.000으로 그 차이는 0.433

In [14]:
# 월별 평균기온
print(df.groupby("Month")['Temp'].mean())

Month
5    65.548387
6    79.100000
7    83.903226
8    83.967742
9    76.900000
Name: Temp, dtype: float64


### 근무자 정보를 측정한 데이터
1. Attrition은 퇴사 여부 데이터를 (yes:1, no:0)으로 데이터를 변환하여 새로운 파생변수에 대입하고 범주별 레코드 수를 계산하시오
2. 범주형 변수 중 유일한 값을 1개만 가지고 있는 컬럼을 찾아내어 그 컬럼을 데이터에서 제거하시오
3. 원래 데이터에서 숫자형인 컬럼만 추출하여 새로운 데이터프레임으로 생성하고, 각 변수간의 상관계수(피어슨 상관계수)를 구하시오

In [15]:
# 필요한 파이썬 라이브러리 불러오기
import numpy as np
import pandas as pd
import seaborn as sns

In [17]:
# csv 파일이 위치한 디렉토리 입력
df = pd.read_csv('data/HR-Employee-Attrition.csv')
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [19]:
## 퇴사 여부 데이터를 (yes:1, no:0)으로 데이터를 변환하여 새로운 파생변수에 대입하고 범주별 레코드 수를 계산
target_map = {'Yes':1, 'No':0}

df["Attrition_numerical"] = df["Attrition"].apply(lambda x: target_map[x])
df["Attrition_numerical"].value_counts()

0    1233
1     237
Name: Attrition_numerical, dtype: int64

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [22]:
## 범주형 변수 중 유일한 값을 1개만 가지고 있는 컬럼을 찾아내어 그 컬럼을 데이터에서 제거
cat_feat = df.select_dtypes('object','category').columns.values
df_cat = df[cat_feat].copy()
df_cat.nunique().sort_values()

Over18            1
Attrition         2
Gender            2
OverTime          2
BusinessTravel    3
Department        3
MaritalStatus     3
EducationField    6
JobRole           9
dtype: int64

In [24]:
## 범주형 변수 중 유일한 값을 1개만 가지고 있는 컬럼은 Over18 이다
df_cat = df_cat.drop(['Over18'], axis=1, errors='ignore')

In [39]:
## 숫자형인 컬럼만 추출하여 새로운 데이터프레임으로 생성
num_feat = df.select_dtypes('number').columns
attrition_num = df[num_feat].copy().corr(method='pearson')


In [41]:
attrition_num

Age  DailyRate  DistanceFromHome  Education  \
Age                       1.000000   0.010661         -0.001686   0.208034   
DailyRate                 0.010661   1.000000         -0.004985  -0.016806   
DistanceFromHome         -0.001686  -0.004985          1.000000   0.021042   
Education                 0.208034  -0.016806          0.021042   1.000000   
EmployeeCount                  NaN        NaN               NaN        NaN   
EmployeeNumber           -0.010145  -0.050990          0.032916   0.042070   
EnvironmentSatisfaction   0.010146   0.018355         -0.016075  -0.027128   
HourlyRate                0.024287   0.023381          0.031131   0.016775   
JobInvolvement            0.029820   0.046135          0.008783   0.042438   
JobLevel                  0.509604   0.002966          0.005303   0.101589   
JobSatisfaction          -0.004892   0.030571         -0.003669  -0.011296   
MonthlyIncome             0.497855   0.007707         -0.017014   0.094961   
MonthlyRate               0.028051  -0.032182          0.027473  -0.026084   
NumCompaniesWorked        0.299635   0.038153         -0.029251   0.126317   
PercentSalaryHike         0.003634   0.022704          0.040235  -0.011111   
PerformanceRating         0.001904   0.000473          0.027110  -0.024539   
RelationshipSatisfaction  0.053535   0.007846          0.006557  -0.009118   
StandardHours                  NaN        NaN               NaN        NaN   
StockOptionLevel          0.037510   0.042143          0.044872   0.018422   
TotalWorkingYears         0.680381   0.014515          0.004628   0.148280   
TrainingTimesLastYear    -0.019621   0.002453         -0.036942  -0.025100   
WorkLifeBalance          -0.021490  -0.037848         -0.026556   0.009819   
YearsAtCompany            0.311309  -0.034055          0.009508   0.069114   
YearsInCurrentRole        0.212901   0.009932          0.018845   0.060236   
YearsSinceLastPromotion   0.216513  -0.033229          0.010029   0.054254   
YearsWithCurrManager      0.202089  -0.026363          0.014406   0.069065   
Attrition_numerical      -0.159205  -0.056652          0.077924  -0.031373   

                          EmployeeCount  EmployeeNumber  \
Age                                 NaN       -0.010145   
DailyRate                           NaN       -0.050990   
DistanceFromHome                    NaN        0.032916   
Education                           NaN        0.042070   
EmployeeCount                       NaN             NaN   
EmployeeNumber                      NaN        1.000000   
EnvironmentSatisfaction             NaN        0.017621   
HourlyRate                          NaN        0.035179   
JobInvolvement                      NaN       -0.006888   
JobLevel                            NaN       -0.018519   
JobSatisfaction                     NaN       -0.046247   
MonthlyIncome                       NaN       -0.014829   
MonthlyRate                         NaN        0.012648   
NumCompaniesWorked                  NaN       -0.001251   
PercentSalaryHike                   NaN       -0.012944   
PerformanceRating                   NaN       -0.020359   
RelationshipSatisfaction            NaN       -0.069861   
StandardHours                       NaN             NaN   
StockOptionLevel                    NaN        0.062227   
TotalWorkingYears                   NaN       -0.014365   
TrainingTimesLastYear               NaN        0.023603   
WorkLifeBalance                     NaN        0.010309   
YearsAtCompany                      NaN       -0.011240   
YearsInCurrentRole                  NaN       -0.008416   
YearsSinceLastPromotion             NaN       -0.009019   
YearsWithCurrManager                NaN       -0.009197   
Attrition_numerical                 NaN       -0.010577   

                          EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
Age                                      0.010146    0.024287        0.029820   
DailyRate                      